feat(har): Remotely accessible HAR file

rwoll opened the pull request:
This change ensure's the HAR file is saved at `recordHar.path` on the
client instead of the server.

NB: The goal was to make this change transparent to the user and NOT
introduce any new APIs. Namely, I want to leave the API open for
potential `context.har.start()` and `context.har.stop()`.

This does BREAK servers that expect the HAR to be at the `recordHar.path`
on the server, but I think that's OK since there haven't been reports
of missing HAR on client making me think not many users are getting
HAR with client and server on different hosts anyways.

Closes #8355

rwoll commented:
@pavelfeldman / @dgozman: This is ready for review. Thanks!

pavelfeldman commented on src/server/supplements/har/harTracer.ts:
```diff
@@ -263,6 +267,8 @@ export class HarTracer {
   }
 
   async flush() {
+    if (this._isFlushed) return;
```
nit: we put return on the next line

pavelfeldman commented on src/protocol/protocol.yml:
```diff
@@ -684,6 +684,10 @@ BrowserContext:
       returns:
         artifact: Artifact
 
+    harExport:
```
nit: this is ok, but there was nothing wrong with returning optional artifact from close.

pavelfeldman commented on src/server/browserContext.ts:
```diff
@@ -74,7 +74,8 @@ export abstract class BrowserContext extends SdkObject {
 
     if (this._options.recordHar)
-      this._harTracer = new HarTracer(this, this._options.recordHar);
+      this._harTracer = new HarTracer(this, {...this._options.recordHar, path: path.join(this._browser.options.artifactsDir, `${createGuid}.har`)});
```
I don't think this does what you think this does, which suggests we might be missing a test with 2 hars.

rwoll replied:
Oops! Missed the actual invocation!

pavelfeldman reviewed (commented):
Looks good, has a small bug.

rwoll commented:
Thanks @pavelfeldman! It's all fixed up and I added a test for the duplicate/wrong filename.

The test failed before the change (8/10 times with `--repeat-each`), but now passes 🎉! (I couldn't get a more deterministic failure since once the context is closed, we don't call saveAs again.)

rwoll committed:
feat(har): Remotely accessible HAR file

This change ensure's the HAR file is saved at `recordHar.path` on the
client instead of the server.

NB: The goal was to make this change transparent to the user and NOT
introduce any new APIs. Namely, I want to leave the API open for
potential `context.har.start()` and `context.har.stop()`.

This does BREAK servers that expect the HAR to be at the `recordHar.path`
on the server, but I think that's OK since there haven't been reports
of missing HAR on client making me think not many users are getting
HAR with client and server on different hosts anyways.

Closes #8355
